packages/go_router/lib/src/route.dart

```diff
@@ -458,4 +558,771 @@ class ShellRoute extends RouteBase {
+/// own Navigator. A container widget responsible for managing the Navigators
+/// for all route branches will then be passed as the child argument
+/// of the builder function.
+///
+/// ```
+/// final GlobalKey<NavigatorState> _tabANavigatorKey =
+///   GlobalKey<NavigatorState>(debugLabel: 'tabANavigator');
+/// final GlobalKey<NavigatorState> _tabBNavigatorKey =
+///   GlobalKey<NavigatorState>(debugLabel: 'tabBNavigator');
+///
+/// final GoRouter _router = GoRouter(
+///   initialLocation: '/a',
+///   routes: <RouteBase>[
+///     StatefulShellRoute(
+///       builder: (BuildContext context, GoRouterState state,
+///             StatefulNavigationShell navigationShell) {
+///         return ScaffoldWithNavBar(navigationShell: navigationShell);
+///       },
+///       navigatorContainerBuilder: (BuildContext context,
+///             StatefulNavigationShell navigationShell,
+///             List<Widget> children) =>
+///                 MyCustomContainer(children: children),
+///       branches: [
+///         /// The first branch, i.e. tab 'A'
+///         StatefulShellBranch(
+///           navigatorKey: _tabANavigatorKey,
+///           routes: <RouteBase>[
+///             GoRoute(
+///               path: '/a',
+///               builder: (BuildContext context, GoRouterState state) =>
+///                   const RootScreen(label: 'A', detailsPath: '/a/details'),
+///               routes: <RouteBase>[
+///                 /// Will cover screen A but not the bottom navigation bar
+///                 GoRoute(
+///                   path: 'details',
+///                   builder: (BuildContext context, GoRouterState state) =>
+///                       const DetailsScreen(label: 'A'),
+///                 ),
+///               ],
+///             ),
+///           ],
+///         ),
+///         /// The second branch, i.e. tab 'B'
+///         StatefulShellBranch(
+///           navigatorKey: _tabBNavigatorKey,
+///           routes: <RouteBase>[
+///             GoRoute(
+///               path: '/b',
+///               builder: (BuildContext context, GoRouterState state) =>
+///                   const RootScreen(label: 'B', detailsPath: '/b/details'),
+///               routes: <RouteBase>[
+///                 /// Will cover screen B but not the bottom navigation bar
+///                 GoRoute(
+///                   path: 'details',
+///                   builder: (BuildContext context, GoRouterState state) =>
+///                       const DetailsScreen(label: 'B'),
+///                 ),
+///               ],
+///             ),
+///           ],
+///         ),
+///       ],
+///     ),
+///   ],
+/// );
+/// ```
+///
+/// See [Stateful Nested Navigation](https://github.com/flutter/packages/blob/main/packages/go_router/example/lib/stacked_shell_route.dart)
+/// for a complete runnable example using StatefulShellRoute and StatefulShellRoute.
+class StatefulShellRoute extends ShellRouteBase {
```

Comment:
Why does hot reloading reset the state?

Reply:
I just tried running the sample code in macOS and iOS, and hot reload seemed to work as expected. Which environment did you test on?

Reply:
I tested with this app on iOS and Desktop:

```
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';

void main() {
  runApp(MyApp());
}

class Foo extends StatelessWidget {
  const Foo({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const Placeholder();
  }
}


///
class MyApp extends StatelessWidget {
  ///
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp.router(
      routerConfig: GoRouter(
        routes: <RouteBase>[
          StatefulShellRoute.indexedStack(
            branches: [
              StatefulShellBranch(
                routes: [
                  GoRoute(
                    path: '/',
                    pageBuilder: (BuildContext context, GoRouterState state) {
                      return FadeTransitionPage(key: state.pageKey, child: HomeScreen());
                    },
                  ),
                ],
              ),
              StatefulShellBranch(
                routes: [
                  GoRoute(
                    path: '/work',
                    pageBuilder: (BuildContext context, GoRouterState state) {
                      return FadeTransitionPage(key: state.pageKey, child: WorkScreen());
                    },
                  ),
                ],
              ),
            ],
            builder: (BuildContext context, GoRouterState state,
                StatefulNavigationShell navigationShell) {
              return AppScaffold(
                shell: navigationShell,
              );
            },
          ),
        ],
      ),
    );
  }
}

///
class AppScaffold extends StatelessWidget {
  final StatefulNavigationShell shell;

  ///
  const AppScaffold({super.key, required this.shell});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: shell,
      bottomNavigationBar: BottomNavigationBar(
        items: const <BottomNavigationBarItem>[
          BottomNavigationBarItem(icon: Icon(Icons.home), label: 'Home'),
          BottomNavigationBarItem(icon: Icon(Icons.work), label: 'Work'),
        ],
        currentIndex: shell.currentIndex,
        onTap: (int index) => _onTap(context, index),
      ),
    );
  }

  void _onTap(BuildContext context, int index) {
    shell.goBranch(index);
  }
}

///
class HomeScreen extends StatefulWidget {
  ///
  const HomeScreen({super.key});

  @override
  State<HomeScreen> createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> {
  @override
  Widget build(BuildContext context) {
    return const Scaffold(
      body: Center(
        child: Text('Home'),
      ),
    );
  }
}

///
class WorkScreen extends StatefulWidget {
  ///
  const WorkScreen({super.key});

  @override
  State<WorkScreen> createState() => _WorkScreenState();
}

class _WorkScreenState extends State<WorkScreen> {
  int count = 0;
  final Stream<dynamic> _stream = Stream.periodic(Duration(seconds: 1));
  late final StreamSubscription<dynamic> _subscription;

  ///
  void initState() {
    super.initState();
    _subscription = _stream.listen((_) {
      setState(() {
        count++;
      });
    });
  }

  ///
  void dispose() {
    _subscription.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Text('Count!!! $count'),
      ),
    );
  }
}

/// A page that fades in an out.
class FadeTransitionPage extends CustomTransitionPage<void> {
  /// Creates a [FadeTransitionPage].
  FadeTransitionPage({
    required LocalKey super.key,
    required super.child,
  }) : super(
      transitionsBuilder: (BuildContext context,
          Animation<double> animation,
          Animation<double> secondaryAnimation,
          Widget child) =>
          FadeTransition(
            opacity: animation.drive(_curveTween),
            child: child,
          ));

  static final CurveTween _curveTween = CurveTween(curve: Curves.easeIn);
}
```

Reply:
Ok, I'll have a look at your code and see what's going on.

Reply:
Ok, so the main issue with this code I guess is that the GoRouter is recreated every time MyApp is rebuilt. Among other things this means that new and unique Navigators will be created on each hot reload, thus effectively resetting state. 

As I've understood it, the intended way to use GoRouter is by creating a single instance only once, and then passing that instance to MaterialApp.router, like the code below (and like is done in all the sample code and documentation). 
I've certainly run into issues myself (with other routes / ShellRoute) when not doing so.

```dart
final _router = GoRouter(
  routes: [
    ...
  ],
);

Widget build(BuildContext context) {
  return MaterialApp.router(
    routerConfig: _router,
  );
}
```

Reply:
I suspect this is because StatefulNavigationShell doesn't use a `Key`. I don't think users should need to define GoRouter as a field on the library, but if there's Key, the framework won't know which State object to use when the app rebuilds.

Reply:
I agree that possibly this shouldn't be a requirement, but I'm wondering if go_router is built to support this right now? 

Anyway, to make this example work, you would need to use fixed navigatorKeys for both GoRouter and each of the branches. And you're probably (partially) right about the key for StatefulNavigationShell - a GlobalKey is actually used for that Widget though, the problem is that a new key is created each time StatefulShellRoute is instantiated (the shell route holds the key). I tried making it possible to pass this key in the constructor to StatefulShellRoute, but I ended up with new problems with "GlobalKey used by multiple widgets" errors.